-1

Tested at: c5b550b6a4e9cff3084131b227bb0c52e0631ef3

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16393/16170/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  ClangBuild**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
1003.0 step2
<pre>runTheMatrix-results/1003.0_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM/step2_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM.log

</pre>10021.0 step2
<pre>runTheMatrix-results/10021.0_TenMuE_0_200+TenMuE_0_200_pythia8_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step2_TenMuE_0_200+TenMuE_0_200_pythia8_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>136.731 step2
<pre>runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/step2_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2.log

</pre>10024.0 step2
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log

</pre>

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 48 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>/cvmfs/cms-ib.cern.ch/2016-45/slc6_amd64_gcc530/external/gcc/5.3.0/lib/gcc/x86_64-pc-linux-gnu/5.3.0/../../../../include/c++/5.3.0/bits/atomic_base.h:157:26: note: expanded from macro 'ATOMIC_FLAG_INIT'
#define ATOMIC_FLAG_INIT { 0 }
                         ^~~~~
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-11-03-1100/src/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-11-03-1100/src/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc 
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-11-03-1100/src/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc:197:16: error: variable-sized object may not be initialized
  int electron[end-begin] = {0};
               ^~~~~~~~~
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-11-03-1100/src/RecoLocalTracker/SiPixelClusterizer/test/ReadPixClusters.cc:93:16: warning: 'ReadPixClusters::beginRun' hides overloaded virtual function [-Woverloaded-virtual]
  virtual void beginRun(const edm::EventSetup& iSetup);
               ^
</pre>

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/PixelClusterizerBase.h
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/PixelClusterizerBase.h
@@ -16,6 +16,7 @@ class PixelGeomDetUnit;
 class PixelClusterizerBase {
 public:
   typedef edm::DetSet<PixelDigi>::const_iterator    DigiIterator;
+  typedef edmNew::DetSet<SiPixelCluster>::const_iterator    ClusterIterator;
 
   struct AccretionCluster {
     typedef unsigned short UShort;
@@ -53,6 +54,11 @@ public:
 				  const PixelGeomDetUnit * pixDet,
 				  const std::vector<short>& badChannels,
 				  edmNew::DetSetVector<SiPixelCluster>::FastFiller& output) = 0;
+
+  virtual void clusterizeDetUnit( const edmNew::DetSet<SiPixelCluster> & input,
+                                  const PixelGeomDetUnit * pixDet,
+                                  const std::vector<short>& badChannels,
+                                  edmNew::DetSetVector<SiPixelCluster>::FastFiller& output) = 0;
 
   // Configure gain calibration service
   void setSiPixelGainCalibrationService( SiPixelGainCalibrationServiceBase* in){ 

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/PixelClusterizerBase.h
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/PixelClusterizerBase.h
@@ -17,6 +17,33 @@ class PixelClusterizerBase {
 public:
   typedef edm::DetSet<PixelDigi>::const_iterator    DigiIterator;
 
+  struct AccretionCluster {
+    typedef unsigned short UShort;
+    static constexpr UShort MAXSIZE = 256;
+    UShort adc[256];
+    UShort x[256];
+    UShort y[256];
+    UShort xmin=16000;
+    UShort ymin=16000;
+    unsigned int isize=0;
+    unsigned int curr=0;
+
+    // stack interface (unsafe ok for use below)
+    UShort top() const { return curr;}
+    void pop() { ++curr;}
+    bool empty() { return curr==isize;}
+
+    bool add(SiPixelCluster::PixelPos const & p, UShort const iadc) {
+      if (isize==MAXSIZE) return false;
+      xmin=std::min(xmin,(unsigned short)(p.row()));
+      ymin=std::min(ymin,(unsigned short)(p.col()));
+      adc[isize]=iadc;
+      x[isize]=p.row();
+      y[isize++]=p.col();
+      return true;
+    }
+  };
+
   // Virtual destructor, this is a base class.
   virtual ~PixelClusterizerBase() {}
 

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc
@@ -112,13 +112,14 @@ bool PixelThresholdClusterizer::setup(const PixelGeomDetUnit * pixDet)
 //!  each seed pixel.
 //!  Input and output data stored in DetSet
 //----------------------------------------------------------------------------
-void PixelThresholdClusterizer::clusterizeDetUnit( const edm::DetSet<PixelDigi> & input,
+template<typename T>
+void PixelThresholdClusterizer::clusterizeDetUnitT( const T & input,
 						   const PixelGeomDetUnit * pixDet,
 						   const std::vector<short>& badChannels,
                                                    edmNew::DetSetVector<SiPixelCluster>::FastFiller& output) {
   
-  DigiIterator begin = input.begin();
-  DigiIterator end   = input.end();
+  typename T::const_iterator begin = input.begin();
+  typename T::const_iterator end   = input.end();
   
   // Do not bother for empty detectors
   //if (begin == end) cout << " PixelThresholdClusterizer::clusterizeDetUnit - No digis to clusterize";
@@ -180,6 +181,19 @@ void PixelThresholdClusterizer::clear_buffer( DigiIterator begin, DigiIterator e
     }
 }
 
+void PixelThresholdClusterizer::clear_buffer( ClusterIterator begin, ClusterIterator end )
+{
+  for(ClusterIterator ci = begin; ci != end; ++ci )
+    {
+      for(int i = 0; i < ci->size(); ++i)
+        {
+          const SiPixelCluster::Pixel pixel = ci->pixel(i);
+
+          theBuffer.set_adc( pixel.x, pixel.y, 0 );   // reset pixel adc to 0
+        }
+    }
+}
+
 //----------------------------------------------------------------------------
 //! \brief Copy adc counts from PixelDigis into the buffer, identify seeds.
 //----------------------------------------------------------------------------
@@ -237,6 +251,25 @@ void PixelThresholdClusterizer::copy_to_buffer( DigiIterator begin, DigiIterator
   }
   assert(i==(end-begin));
 
+}
+
+void PixelThresholdClusterizer::copy_to_buffer( ClusterIterator begin, ClusterIterator end )
+{
+  // loop over clusters
+  for(ClusterIterator ci = begin; ci != end; ++ci) {
+    // loop over pixels
+    for(int i = 0; i < ci->size(); ++i) {
+      const SiPixelCluster::Pixel pixel = ci->pixel(i);
+
+      int row = pixel.x;
+      int col = pixel.y;
+      int adc = pixel.adc;
+      if ( adc >= thePixelThreshold) {
+        theBuffer.set_adc( row, col, adc);
+        if ( adc >= theSeedThreshold) theSeeds.push_back( SiPixelCluster::PixelPos(row,col) );
+      }
+    }
+  }
 }
 
 //----------------------------------------------------------------------------

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc
@@ -312,38 +312,6 @@ int PixelThresholdClusterizer::calibrate(int adc, int col, int row)
   return electrons;
 }
 
-
-namespace {
-
-  struct AccretionCluster {
-    typedef unsigned short UShort;
-    static constexpr UShort MAXSIZE = 256;
-    UShort adc[256];
-    UShort x[256];
-    UShort y[256];
-    UShort xmin=16000;
-    UShort ymin=16000;
-    unsigned int isize=0;
-    unsigned int curr=0;
-
-    // stack interface (unsafe ok for use below)
-    UShort top() const { return curr;}
-    void pop() { ++curr;}   
-    bool empty() { return curr==isize;}
-
-    bool add(SiPixelCluster::PixelPos const & p, UShort const iadc) {
-      if (isize==MAXSIZE) return false;
-      xmin=std::min(xmin,(unsigned short)(p.row()));
-      ymin=std::min(ymin,(unsigned short)(p.col()));
-      adc[isize]=iadc;
-      x[isize]=p.row();
-      y[isize++]=p.col();
-      return true;
-    }
-  };
-
-}
-
 //----------------------------------------------------------------------------
 //!  \brief The actual clustering algorithm: group the neighboring pixels around the seed.
 //----------------------------------------------------------------------------

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc
@@ -43,23 +43,23 @@ using namespace std;
 //----------------------------------------------------------------------------
 PixelThresholdClusterizer::PixelThresholdClusterizer
   (edm::ParameterSet const& conf) :
-    conf_(conf), bufferAlreadySet(false), theNumOfRows(0), theNumOfCols(0), detid_(0) 
+    bufferAlreadySet(false), theNumOfRows(0), theNumOfCols(0), detid_(0) 
 {
   // Get thresholds in electrons
   thePixelThreshold   = 
-    conf_.getParameter<int>("ChannelThreshold");
+    conf.getParameter<int>("ChannelThreshold");
   theSeedThreshold    = 
-    conf_.getParameter<int>("SeedThreshold");
+    conf.getParameter<int>("SeedThreshold");
   theClusterThreshold = 
-    conf_.getParameter<double>("ClusterThreshold");
+    conf.getParameter<double>("ClusterThreshold");
   theConversionFactor = 
-    conf_.getParameter<int>("VCaltoElectronGain");
+    conf.getParameter<int>("VCaltoElectronGain");
   theOffset = 
-    conf_.getParameter<int>("VCaltoElectronOffset");
-  if ( conf_.exists("AdcFullScaleStack") ) theStackADC_=conf_.getParameter<int>("AdcFullScaleStack");
+    conf.getParameter<int>("VCaltoElectronOffset");
+  if ( conf.exists("AdcFullScaleStack") ) theStackADC_=conf.getParameter<int>("AdcFullScaleStack");
   else 
     theStackADC_=255;
-  if ( conf_.exists("FirstStackLayer") ) theFirstStack_=conf_.getParameter<int>("FirstStackLayer");
+  if ( conf.exists("FirstStackLayer") ) theFirstStack_=conf.getParameter<int>("FirstStackLayer");
   else
     theFirstStack_=5;
   if ( conf_.exists("ElectronPerADCGain") ) theElectronPerADCGain_=conf_.getParameter<double>("ElectronPerADCGain");
@@ -67,7 +67,7 @@ PixelThresholdClusterizer::PixelThresholdClusterizer
     theElectronPerADCGain_=135.;
   
   // Get the constants for the miss-calibration studies
-  doMissCalibrate=conf_.getUntrackedParameter<bool>("MissCalibrate",true); 
+  doMissCalibrate=conf.getUntrackedParameter<bool>("MissCalibrate",true); 
   doSplitClusters = conf.getParameter<bool>("SplitClusters");
   theBuffer.setSize( theNumOfRows, theNumOfCols );
 }

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.h
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.h
@@ -63,11 +63,19 @@ class dso_hidden PixelThresholdClusterizer final : public PixelClusterizerBase {
   void clusterizeDetUnit( const edm::DetSet<PixelDigi> & input,	
 				  const PixelGeomDetUnit * pixDet,
 				  const std::vector<short>& badChannels,
-				  edmNew::DetSetVector<SiPixelCluster>::FastFiller& output
-);
+				  edmNew::DetSetVector<SiPixelCluster>::FastFiller& output) { clusterizeDetUnitT(input, pixDet, badChannels, output); }
+  void clusterizeDetUnit( const edmNew::DetSet<SiPixelCluster> & input,
+                          const PixelGeomDetUnit * pixDet,
+                          const std::vector<short>& badChannels,
+                          edmNew::DetSetVector<SiPixelCluster>::FastFiller& output) { clusterizeDetUnitT(input, pixDet, badChannels, output); }
 
-  
  private:
+
+  template<typename T>
+  void clusterizeDetUnitT( const T & input,
+                           const PixelGeomDetUnit * pixDet,
+                           const std::vector<short>& badChannels,
+                           edmNew::DetSetVector<SiPixelCluster>::FastFiller& output);
 
   //! Data storage
   SiPixelArrayBuffer               theBuffer;         // internal nrow * ncol matrix
@@ -96,7 +104,9 @@ class dso_hidden PixelThresholdClusterizer final : public PixelClusterizerBase {
   //! Private helper methods:
   bool setup(const PixelGeomDetUnit * pixDet);
   void copy_to_buffer( DigiIterator begin, DigiIterator end );   
-  void clear_buffer( DigiIterator begin, DigiIterator end );   
+  void copy_to_buffer( ClusterIterator begin, ClusterIterator end );
+  void clear_buffer( DigiIterator begin, DigiIterator end );
+  void clear_buffer( ClusterIterator begin, ClusterIterator end );
   SiPixelCluster make_cluster( const SiPixelCluster::PixelPos& pix, edmNew::DetSetVector<SiPixelCluster>::FastFiller& output
 );
   // Calibrate the ADC charge to electrons 

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.h
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.h
@@ -69,8 +69,6 @@ class dso_hidden PixelThresholdClusterizer final : public PixelClusterizerBase {
   
  private:
 
-  edm::ParameterSet conf_;
-
   //! Data storage
   SiPixelArrayBuffer               theBuffer;         // internal nrow * ncol matrix
   bool                             bufferAlreadySet;  // status of the buffer array

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.cc
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.cc
@@ -48,7 +48,6 @@
   //---------------------------------------------------------------------------
   SiPixelClusterProducer::SiPixelClusterProducer(edm::ParameterSet const& conf) 
     : 
-    conf_(conf),
     theSiPixelGainCalibration_(0), 
     clusterMode_("None"),     // bogus
     clusterizer_(0),          // the default, in case we fail to make one
@@ -56,6 +55,8 @@
     src_( conf.getParameter<edm::InputTag>( "src" ) ),
     maxTotalClusters_( conf.getParameter<int32_t>( "maxNumberOfClusters" ) )
   {
+    clusterMode_ = conf.getUntrackedParameter<std::string>("ClusterMode","PixelThresholdClusterizer");
+
     tPixelDigi = consumes<edm::DetSetVector<PixelDigi>>(src_);
     //--- Declare to the EDM what kind of collections we will be making.
     produces<SiPixelClusterCollectionNew>(); 
@@ -71,7 +72,7 @@
 
     //--- Make the algorithm(s) according to what the user specified
     //--- in the ParameterSet.
-    setupClusterizer();
+    setupClusterizer(conf);
 
   }
 
@@ -119,12 +120,10 @@
   //!  TO DO: in the future, we should allow for a different algorithm for 
   //!  each detector subset (e.g. barrel vs forward, per layer, etc).
   //---------------------------------------------------------------------------
-  void SiPixelClusterProducer::setupClusterizer()  {
-    clusterMode_ = 
-      conf_.getUntrackedParameter<std::string>("ClusterMode","PixelThresholdClusterizer");
+  void SiPixelClusterProducer::setupClusterizer(const edm::ParameterSet& conf)  {
 
     if ( clusterMode_ == "PixelThresholdClusterizer" ) {
-      clusterizer_ = new PixelThresholdClusterizer(conf_);
+      clusterizer_ = new PixelThresholdClusterizer(conf);
       clusterizer_->setSiPixelGainCalibrationService(theSiPixelGainCalibration_);
       readyToCluster_ = true;
     } 

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.cc
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.cc
@@ -52,12 +52,14 @@
     clusterMode_("None"),     // bogus
     clusterizer_(0),          // the default, in case we fail to make one
     readyToCluster_(false),   // since we obviously aren't
-    src_( conf.getParameter<edm::InputTag>( "src" ) ),
     maxTotalClusters_( conf.getParameter<int32_t>( "maxNumberOfClusters" ) )
   {
     clusterMode_ = conf.getUntrackedParameter<std::string>("ClusterMode","PixelThresholdClusterizer");
 
-    tPixelDigi = consumes<edm::DetSetVector<PixelDigi>>(src_);
+    if ( clusterMode_ == "PixelThresholdReclusterizer" )
+      tPixelClusters = consumes<SiPixelClusterCollectionNew>( conf.getParameter<edm::InputTag>("src") );
+    else
+      tPixelDigi = consumes<edm::DetSetVector<PixelDigi>>( conf.getParameter<edm::InputTag>("src") );
     //--- Declare to the EDM what kind of collections we will be making.
     produces<SiPixelClusterCollectionNew>(); 
 
@@ -92,10 +94,13 @@
     //Setup gain calibration service
     theSiPixelGainCalibration_->setESObjects( es );
 
-   // Step A.1: get input data
-    //edm::Handle<PixelDigiCollection> pixDigis;
-    edm::Handle< edm::DetSetVector<PixelDigi> >  input;
-    e.getByToken(tPixelDigi, input);
+    // Step A.1: get input data
+    edm::Handle< SiPixelClusterCollectionNew >   inputClusters;
+    edm::Handle< edm::DetSetVector<PixelDigi> >  inputDigi;
+    if ( clusterMode_ == "PixelThresholdReclusterizer" )
+      e.getByToken(tPixelClusters, inputClusters);
+    else
+      e.getByToken(tPixelDigi, inputDigi);
 
     // Step A.2: get event setup
     edm::ESHandle<TrackerGeometry> geom;
@@ -107,7 +112,10 @@
 
     // Step C: Iterate over DetIds and invoke the pixel clusterizer algorithm
     // on each DetUnit
-    run(*input, geom, *output );
+    if ( clusterMode_ == "PixelThresholdReclusterizer" )
+      run(*inputClusters, geom, *output );
+    else
+      run(*inputDigi, geom, *output );
 
     // Step D: write output to file
     output->shrink_to_fit();
@@ -122,7 +130,7 @@
   //---------------------------------------------------------------------------
   void SiPixelClusterProducer::setupClusterizer(const edm::ParameterSet& conf)  {
 
-    if ( clusterMode_ == "PixelThresholdClusterizer" ) {
+    if ( clusterMode_ == "PixelThresholdReclusterizer" || clusterMode_ == "PixelThresholdClusterizer" ) {
       clusterizer_ = new PixelThresholdClusterizer(conf);
       clusterizer_->setSiPixelGainCalibrationService(theSiPixelGainCalibration_);
       readyToCluster_ = true;
@@ -136,11 +144,13 @@
     }
   }
 
+
   //---------------------------------------------------------------------------
   //!  Iterate over DetUnits, and invoke the PixelClusterizer on each.
   //---------------------------------------------------------------------------
-  void SiPixelClusterProducer::run(const edm::DetSetVector<PixelDigi>   & input, 
-				   edm::ESHandle<TrackerGeometry>       & geom,
+  template<typename T>
+  void SiPixelClusterProducer::run(const T                              & input, 
+                                   const edm::ESHandle<TrackerGeometry> & geom,
                                    edmNew::DetSetVector<SiPixelCluster> & output) {
     if ( ! readyToCluster_ ) {
       edm::LogError("SiPixelClusterProducer")
@@ -153,7 +163,7 @@
     int numberOfClusters = 0;
  
     // Iterate on detector units
-    edm::DetSetVector<PixelDigi>::const_iterator DSViter = input.begin();
+    typename T::const_iterator DSViter = input.begin();
     for( ; DSViter != input.end(); DSViter++) {
       ++numberOfDetUnits;
 

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.h
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.h
@@ -61,18 +61,19 @@
     virtual void produce(edm::Event& e, const edm::EventSetup& c) override;
 
     //--- Execute the algorithm(s).
-    void run(const edm::DetSetVector<PixelDigi>   & input,
-	     edm::ESHandle<TrackerGeometry>       & geom,
+    template<typename T>
+    void run(const T                              & input,
+             const edm::ESHandle<TrackerGeometry> & geom,
              edmNew::DetSetVector<SiPixelCluster> & output);
 
   private:
+    edm::EDGetTokenT<SiPixelClusterCollectionNew>  tPixelClusters;
     edm::EDGetTokenT<edm::DetSetVector<PixelDigi>> tPixelDigi;
     // TO DO: maybe allow a map of pointers?
     SiPixelGainCalibrationServiceBase * theSiPixelGainCalibration_;
     std::string clusterMode_;               // user's choice of the clusterizer
     PixelClusterizerBase * clusterizer_;    // what we got (for now, one ptr to base class)
     bool readyToCluster_;                   // needed clusterizers valid => good to go!
-    edm::InputTag src_;
 
     //! Optional limit on the total number of clusters
     int32_t maxTotalClusters_;

--- a/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.h
+++ b/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.h
@@ -55,7 +55,7 @@
     explicit SiPixelClusterProducer(const edm::ParameterSet& conf);
     virtual ~SiPixelClusterProducer();
 
-    void setupClusterizer();
+    void setupClusterizer(const edm::ParameterSet& conf);
 
     //--- The top-level event method.
     virtual void produce(edm::Event& e, const edm::EventSetup& c) override;
@@ -66,7 +66,6 @@
              edmNew::DetSetVector<SiPixelCluster> & output);
 
   private:
-    edm::ParameterSet conf_;
     edm::EDGetTokenT<edm::DetSetVector<PixelDigi>> tPixelDigi;
     // TO DO: maybe allow a map of pointers?
     SiPixelGainCalibrationServiceBase * theSiPixelGainCalibration_;

